# Variables
* we can use python like a calculator

3 + 5 * 4

* or we can assign things to variables

In [2]:
weight_kg = 60

* weight_kg is variable name
* variable names are case sensitive, can include letters, digits and underscores
* can't start with a digit

* python has different data types
* integers, floats, strings
* automatically determined, 60 is an integer, 60.3 is a float

In [3]:
weight_kg = 60.3

* strings are surrouneded by single or double quotes

In [4]:
patient_id = '001'

* we can store the results of calculations in a variable

In [5]:
weight_lb = 2.2 * weight_kg

* we can add a prefix to our patient id, this joins two strings together

In [6]:
patient_id = 'inflam_' + patient_id

* functions do something with some data
* lots of built in functions like print
* brackets around parameters to the function

In [7]:
print(weight_lb)

132.66


In [8]:
print(patient_id)

inflam_001


* display multiple things in one call to print by using commas
* automatically adds spaces

In [9]:
print(patient_id, 'weight in kilos:', weight_kg)

inflam_001 weight in kilos: 60.3


* function calls can be nested
* type function tells us what type a variable is

In [10]:
print(type(60.3))

<class 'float'>


In [11]:
print(type(patient_id))

<class 'str'>


* we can do arithmetic inside a function call

In [12]:
print('weight in pounds:', 2.2 * weight_kg)

weight in pounds: 132.66


In [13]:
print(weight_kg)

60.3


* we didn't change weight_kg
* have to assign a new value with = 

In [14]:
weight_kg = 65.0
print('weight in kilos is now:', weight_kg)

weight in kilos is now: 65.0


In [15]:
weight_lb = 2.2 * weight_kg
print('weight in kilos:', weight_kg, 'and in pounds:', weight_lb)

weight in kilos: 65.0 and in pounds: 143.0


In [16]:
weight_kg = 100.0
print('weight in kilos is now:', weight_kg, 'and weight in pounds is still:', weight_lb)

weight in kilos is now: 100.0 and weight in pounds is still: 143.0


* think of variables like sticky notes
* each has a name and value
* changing one doesn't change others

## Exercises

In [17]:
mass = 47.5
age = 122
mass = mass * 2.0
age = age - 20



In [18]:
first, second = 'Grace', 'Hopper'
third, fourth = second, first
print(third, fourth)



Hopper Grace


In [19]:
planet = 'Earth'
apples = 5
distance = 10.5



## recap

* Basic data types in Python include integers, strings, and floating-point numbers.
* Use variable = value to assign a value to a variable in order to record it in memory.
* Variables are created on demand whenever a value is assigned to them.
* Use print(something) to display the value of something.
* Built-in functions are always available to use.


# Analysing some wave-height data

In [2]:
import numpy

* explain libraries, like taking equipment out of storage.
* use import statement
* libraryname.function name syntax

In [3]:
numpy.loadtxt(fname='data/wavesmonthly.csv', delimiter=',', skiprows=1)


array([[1.979e+03, 1.000e+00, 3.788e+00],
       [1.979e+03, 2.000e+00, 3.768e+00],
       [1.979e+03, 3.000e+00, 4.774e+00],
       ...,
       [2.015e+03, 1.000e+01, 3.046e+00],
       [2.015e+03, 1.100e+01, 4.622e+00],
       [2.015e+03, 1.200e+01, 5.048e+00]])

* data now loaded, only subset onscreen
* explain parameters
* skipped first row since it contains headers
* we didn't save the data to a variable, correct that with variable assignment

In [5]:
data = numpy.loadtxt(fname='data/wavesmonthly.csv', delimiter=',', skiprows=1)

In [6]:
print(data)

[[1.979e+03 1.000e+00 3.788e+00]
 [1.979e+03 2.000e+00 3.768e+00]
 [1.979e+03 3.000e+00 4.774e+00]
 ...
 [2.015e+03 1.000e+01 3.046e+00]
 [2.015e+03 1.100e+01 4.622e+00]
 [2.015e+03 1.200e+01 5.048e+00]]


In [7]:
print(type(data))

<class 'numpy.ndarray'>


* numpy arrays can contain multiple data types
* the type function tells us the type of a variable
* dtype tells us the type numpy is using within an array


In [8]:
print(data.dtype)

float64


* find out the array's shape

In [9]:
print(data.shape)

(444, 3)


* that's 444 rows, 3 columns
* 37 years x 12 months = 444
* 3 columns are year, month, waveheight
* get data from an array with [x,y]

In [10]:
print('first value in data:', data[0, 2])

first value in data: 3.788


In [11]:
print("middle value in data:", data[222, 2])

middle value in data: 2.446


* show diagram in notes about arrays
* arrays are zero based indexing
* 0,0 is top left corner, matrix notation not cartesian

## slicing data
* can select multiple items in an array with a :

In [12]:
print(data[0:12, 0:3])

[[1.979e+03 1.000e+00 3.788e+00]
 [1.979e+03 2.000e+00 3.768e+00]
 [1.979e+03 3.000e+00 4.774e+00]
 [1.979e+03 4.000e+00 2.818e+00]
 [1.979e+03 5.000e+00 2.734e+00]
 [1.979e+03 6.000e+00 2.086e+00]
 [1.979e+03 7.000e+00 2.066e+00]
 [1.979e+03 8.000e+00 2.236e+00]
 [1.979e+03 9.000e+00 3.322e+00]
 [1.979e+03 1.000e+01 3.512e+00]
 [1.979e+03 1.100e+01 4.348e+00]
 [1.979e+03 1.200e+01 4.628e+00]]


* 0:12 means from zero to 11
* takes some getting used to, I get it wrong all the time
* slices can start from non-zero positions

In [13]:
print(data[12:24, 1:3])

[[ 1.     3.666]
 [ 2.     4.326]
 [ 3.     3.522]
 [ 4.     3.18 ]
 [ 5.     1.954]
 [ 6.     1.72 ]
 [ 7.     1.86 ]
 [ 8.     1.95 ]
 [ 9.     3.11 ]
 [10.     3.78 ]
 [11.     3.474]
 [12.     5.28 ]]


* we can miss out the start or end if we want everything up to or everything after an item

In [14]:
first_year = data[:12, 2:]
print('data from first year is:')
print(first_year)


data from first year is:
[[3.788]
 [3.768]
 [4.774]
 [2.818]
 [2.734]
 [2.086]
 [2.066]
 [2.236]
 [3.322]
 [3.512]
 [4.348]
 [4.628]]


* numpy has functions which can operate on whole arrays
* mean is one of these

In [15]:
print(numpy.mean(data))

668.9611876876877


* mean takes an argument (data)
* we can also slice data to just get the mean of a subset of the data

In [16]:
print(numpy.mean(data[:,2]))

3.383563063063063


* Not all functions take an argument

In [17]:
import time
print(time.ctime())



Tue May 16 22:10:08 2023


* ctime from the time library needs no arguments, still need the brackets
* lets combine multiple assignment, numpy's min/max/std functions and slicing

In [18]:
maxval, minval, stdval = numpy.max(data[:,2]), numpy.min(data[:,2]), numpy.std(data[:,2])

print('Max wave height:', maxval)
print('Min wave height:', minval)
print('Wave height standard deviation:', stdval)


Max wave height: 6.956
Min wave height: 1.496
Wave height standard deviation: 1.1440155050316319


* don't use variable names with the same name as functions (e.g. min/max)
* mystery functions